dhis2/dhis2-core
# feat: metadata filter for attribute properties [DHIS2-13158]

### Summary 
Extends the metadata API to allows use of `filter` parameters of the form

    filter={attribute-id}:{operator}:{attribute-value}

### Implementation Note
The initial idea to rewrite the expression in the parsing or planning phase of query processing turned out to be unsatisfactory. 
Each filter would become a sub-conjunction of 2 restrictions which made the code and processing quite complicated yet due to the way collection properties like the attributes are processed it would not actually check that both conditions were true for the same attribute but for any of the present attributes which is not what we want. 

Therefore the attribute case is handled explicitly. A path not found in the schema ending with a valid UID is assumed to be an attribute value filter. This is made explicit in the `Restriction` by the added `boolean` field `attribute`.
When in memory filters are processed the value resolution is then adjusted to recognize this and get the attribute value.

### Automatic Testing
New controller test was added. 
I added a new class for filter controller tests of the metadata API.
So far there aren't further explicit tests of filters. This is the start of something that hopefully gets extended in the future.

### Manual Testing
* open maintenance app
* edit an existing attribute to allow it being used for a particular metadata object, for example _additional name_ for organisation units
* edit one or more metadata objects and provide a value for the chosen attribute
* filter based on the attribute uid, for example`/api/organisationUnits?filter={attr-id}:eq:test`
* check the filter results find the metadata objects that should match the filter given they have a matching attribute value
